### ISSUE TITLE
Flux Cirkwi : évolutions à venir

### ISSUE THREAD

**Issue by babastienne** [GeotrekCE/Geotrek-admin#3947]
Le département 62 a financé le développement d'évolutions relatives à Cirkwi et Geotrek.

Le CD62 exporte déjà certaines données vers Cirkwi. Toutefois, il souhaite également importer des itinéraires et POI provenant de Cirkwi dans Geotrek (mode sans segmentation dynamique).

Les développements à venir seront donc les suivants : 
- Ajout d'un parser permettant d'intégrer dans Geotrek des itinéraires provenant de Cirkwi (mode NDS only). [Modèle du flux utilisé](https://ws.cirkwi.com/tuto/flux_circuits/)
- Ajout d'un parser permettant d'intégrer dans Geotrek les POI provenant de Cirkwi. [Modèle du flux](https://ws.cirkwi.com/tuto/flux_pois/)
- Ajout d'un filtre sur l'API Cirkwi de Geotrek, qui permettra de choisir si on souhaite filtrer les données afin de ne pas renvoyer les itinéraires ou POI qui proviennent d'autres flux. Concrètement une méthode regardera pour les modèles si les objets ont des `eid` et si oui le filtre ne renverra pas les données, afin d'éviter des boucles d'import infinies.

Les développements ont étés spécifiés en accord avec Cirkwi et devraient démarrer dans les prochaines semaines.

**Comment by babastienne** [comment]
L'idée du filtre était aussi d'alléger le contenu des réponses du serveur pour avoir moins de résultats et pas renvoyer des itinéraires complets qui ne seront pas exploités. 

C'était un choix de développement : soit on continue d'envoyer de grosses requêtes en ajoutant en plus du reste le contenu du champ provider, soit on pré-filtre en limitant les résultats. Ca aurait été sur une autre API j'aurai plutôt dis non mais là comme dans tous les cas c'est une API spécifique pour Cirkwi ça ne me choque pas d'ajouter ce comportement.

**Comment by camillemonchicourt** [comment]
Oui je parlais d'ajouter un filtre par "Provider" pas de retourner tout et de filtrer après.

Mais là il faudrait exclure un Provider donc certainement plus complexe.

Je me demandais juste si il ne pourrait pas y avoir des contenus ayant un EID mais venant d'ailleurs que Cirkwi et qu'on voudrait donc leur envoyer, mais peut-être que ce n'est pas le cas dans votre contexte.

Selon moi exclure un Provider serait plus propre et plus solide mais peut-être plus complexe et pas utile dans votre contexte.

**Comment by babastienne** [comment]
Ah oui ok je comprends mieux. Pertinent en effet, à réfléchir. Avoir la possibilité peut-être de paramétrer dans le custom.py un ou plusieurs providers qui serviraient de filtres ajoute une plus grande personnalisation, mais à voir en terme de complexité aussi.

Là l'usage qui avait été identifié c'était de ne retourner aucun contenu venant d'une source externe (cirkwi mais potentiellement aussi apidae, tourinsoft, etc.) pour ne conserver que les contenus d'origine Geotrek mais peut-être qu'on pourrait rendre le développement plus générique. J'essayerai d'en discuter avec cirkwi pour voir s'ils imaginent des cas d'usages où ils seraient intéressés pour récupérer des contenus venant d'autres sources ou s'ils ont déjà des passerelles avec ces sources.

**Comment by camillemonchicourt** [comment]
Filtre `?include_externals=false` ajouté à l'API Cirkwi dans la version 2.102.2.

**Comment by camillemonchicourt** [comment]
Ça m'interroge d'avoir des parsers dans Geotrek qui ne peuvent alimenter que des Geotrek sans segmentation dynamique (qui est un usage marginal et qu'il faut limiter au maximum).

Ou du moins il faut être très explicite dans la doc et le code que ceux-ci ne sont utilisables que dans des cas particuliers, pour ne pas que des vrais Geotrek normaux tentent de l'utiliser et ne comprennent pas pourquoi ça ne fonctionne pas.

**Comment by babastienne** [comment]
Oui ok pour le mentionner explicitement et le rendre bien visible dans la doc en effet. :+1: 

Ce n'est pas le premier parser qui existe uniquement pour le mode sans segmentation dynamique. Techniquement c'est aussi le cas de tous les parsers liés au Geotrek Agrégateur. C'est aussi le cas pour le parser d'itinéraires Tourinsoft qui avait été réalisé pour le CD64 mais pas versionné dans le code du fait de la structure changeante des données Tourinsoft.

**Comment by camillemonchicourt** [comment]
Ouais OK, à bien préciser en effet. 
Pour le parser de treks Tourinsoft CD64 pour moi il alimentait bien un Geotrek normal mais peut-être que je m'embrouille.

**Comment by babastienne** [comment]
Ah oui bien vu, c'est moi qui m'embrouille. Il alimente un Geotrek normal mais ne permet donc pas d'importer les traces des itinéraires (dommage).

**Comment by AudreyRemy** [comment]
Je confirme nous avons bien un Geotrek avec segmentation dynamique mais la géométrie n'est pas importée, il est nécessaire de réaliser le tracé dans Geotrek avant publication de l'itinéraire

**Comment by camillemonchicourt** [comment]
Oui, d'ailleurs on avait commencé à en discuter dans le ticket sur le sujet, mais plus globalement, il y a des pistes qu'il serait intéressant d'investiguer pour aussi pouvoir importer aussi les géométries (même à minima) des itinéraires importées, même dans un Geotrek classique.
Gros sujet à étudier et discuter dans un ticket dédié.

**Comment by babastienne** [comment]
La fonctionnalité d'import d'itinéraires et de contenus touristiques depuis Cirkwi a été développée et intégrée dans le [version 2.111.0](https://github.com/GeotrekCE/Geotrek-admin/releases/tag/2.111.0) de Geotrek-Admin.

Par défaut les contenus intégrés sont automatiquement publiés.

Pour pouvoir intégrer ces données, il faut modifier le fichier `parsers.py` pour créer un parser dédié et interroger un flux fourni par Cirkwi. 

**Structure des classes à intégrer dans le `parsers.py`**

```python
class ImportItinerairesCirkwi(CirkwiTrekParser):
    url = "<URL du flux de données Itinéraires"  # Sous la forme https://ws.cirkwi.com/flux/<user>/<code>/circuits.php?widget-id=<id>
    user = "<Nom d'utilisateur>"
    password = "<Mot de passe>"
    auth = (user, password)
    label = "Itinéraires Cirkwi"
    delete = True
    create = True
    provider = "Cirkwi"


class ImportContenusTouristiquesCirkwi(CirkwiTouristicContentParser):
    url = "<URL du flux de données Itinéraires"  # Sous la forme https://ws.cirkwi.com/flux/<user>/<code>/circuits.php?widget-id=<id>"
    user = "<Nom d'utilisateur>"
    password = "<Mot de passe>"
    auth = (user, password)
    label = "Contenus Touristiques Cirkwi"
    delete = True
    create = True
    provider = "Cirkwi"
    # results_path = "circuit/pois/poi"  # Décommenter cette ligne si les contenus touristiques à importer doivent provenir du même flux que les itinéraires
```

Merci au CD62 pour le financement de cette contribution et à @Chatewgne pour le développement :pray: :tada:
